Fix windows log file rotation with readers

cpuguy83 opened the pull request:
This fixes the case where log rotation fails on Windows while there are
clients reading container logs.

Fixes https://github.com/moby/moby/issues/40999

cpuguy83 committed:
Fix windows log file rotation with readers

This fixes the case where log rotation fails on Windows while there are
clients reading container logs.

Evicts readers if there is an error during rotation and try rotation again.
This is needed for Windows with this scenario:

1. `docker logs -f` is called
2. Log rotation occurs (log.txt -> log.txt.1, truncate and re-open
   log.txt)
3. Log rotation occurs again (rm log.txt.1, log.txt -> log.txt.1)

On step 3, before this change, the log rotation will fail with `Access
is denied`.
In this case, what we have is a reader holding a file handle to the
primary log file. The log file is then rotated, but the reader still has
a the handle open. `FILE_SHARE_DELETE` allows this to happen... but then
we try to do it again for the next rotation and it blows up.
So when it blows up we force all the readers to disconnect, close the
log file, and try rotation again, which will succeed based on the added
tests.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah commented:
Seeing some failures on Windows RS5;

```
=== RUN   TestReadLog/tail_exact
    --- FAIL: TestReadLog/tail_exact (0.00s)
        local_test.go:146: assertion failed: error is not nil: error opening rotated log file: error opening file: The system cannot find the file specified.

=== RUN   TestReadLog/tail_less_than_available
    --- FAIL: TestReadLog/tail_less_than_available (0.00s)
        local_test.go:156: assertion failed: error is not nil: error opening rotated log file: error opening file: The system cannot find the file specified.

=== RUN   TestReadLog/tail_more_than_available
    --- FAIL: TestReadLog/tail_more_than_available (0.00s)
        local_test.go:164: assertion failed: error is not nil: error opening rotated log file: error opening file: The system cannot find the file specified.
```

cpuguy83 commented:
I see the problem. It's an error handling issue with `openFile`

cpuguy83 commented:
Should be fixed now.

thaJeztah commented:
Hmm.. one failure remaining on RS5;

```
=== RUN   TestCheckCapacityAndRotate
--- FAIL: TestCheckCapacityAndRotate (0.01s)
    logfile_test.go:294: assertion failed: error is not nil: rename C:\Users\ContainerAdministrator\AppData\Local\Temp\TestCheckCapacityAndRotate612011099\log214931454 C:\Users\ContainerAdministrator\AppData\Local\Temp\TestCheckCapacityAndRotate612011099\log214931454.1: Access is denied.
```

cpuguy83 commented:
I do not know why this is failing now... It didn't fail until I changed to not wrap the error from open...

cpuguy83 commented:
Ok, so I had to change the this to evict log readers when there is an error rotating files.
I found that in this scenario:

1. Client opens log stream
2. Rotation occurs
3. Rotation occurs again

On step 3, the rotation will fail when it is trying to rename the current file.
It will either get an "Access is denied" error, or "file is being used by another process" error.

So on step 3, when an error occurs, it will evict readers and try again (a few times until successful || max retries).

The thinking here is allowing the container to make progress is more important than the log reader reading the logs.
We can't really seamlessly transfer the log reader without dropping logs anyway, so take the simpler route and have the client reconnect.
We pass the eviction error along to the client.

Without this change, log rotation is essentially broken on Windows if there are active log consumers.

cpuguy83 committed:
Fix various race conditions in loggerutils

Found by running with `go test -race`

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
Added a 2nd commit to fix some race conditions in the loggerutils tests.
Additionally, I think I fixed the flakey test. See commit for details.

cpuguy83 committed:
Clean up some logfile implementation

- Ignore some pointless errors (like not exist on remove)
- Consolidate error handling/logging
- Fix race condition reading last log timestamp in the compression
  goroutine. This needs to be done while holding the write lock, which
  is not (or may not be) locked while compressing a rotated log file.
- Remove some indentation and consolidate mutex unlocking in
  `compressFile`

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

thaJeztah commented:
@kevpar @kolyshkin ptal

cpuguy83 committed:
Fix logfile to open all files with custom openFile

This makes sure, on Windows, that all files are opened with
FILE_SHARE_DELETE.

On non-Windows this just calls the same `os.Open()`.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
And split up commits for easier review.

cpuguy83 commented:
All green.

thaJeztah commented:
@kevpar @kolyshkin ptal

kolyshkin commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -204,15 +204,31 @@ func (w *LogFile) checkCapacityAndRotate() (retErr error) {
 
 	if err := rotate(fname, w.maxFiles, w.compress); err != nil {
-		return err
+		logrus.WithError(err).Warn("Error rotating log file, log data may hav been lost")
```
nit: `s/hav/have/`

kolyshkin commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -569,7 +605,7 @@ func tailFiles(files []SizeReaderAt, watcher *logger.LogWatcher, dec Decoder, ge
 }
 
-func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan interface{}, dec Decoder, since, until time.Time) {
+func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate, notifytEvict chan interface{}, dec Decoder, since, until time.Time) {
```
nit: `s/notifytEvict/notifyEvict/`

thaJeztah commented:
@kolyshkin nits are addressed; PTAL

thaJeztah commented:
@kevpar @kolyshkin @tonistiigi ptal

kevpar commented:
I don't have much context on this code, but the issue I'm aware of should be fixed by switching to do file opens with `FILE_SHARE_DELETE`. It looks like this PR consistently uses the `open`/`openFile` wrappers which do so, so LGTM.

thaJeztah commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -179,45 +179,54 @@ func (w *LogFile) WriteLogEntry(msg *logger.Message) error {
+	w.currentSize = 0
+	w.notifyRotate.Publish(struct{}{})
+
+	if w.maxFiles <= 1 || !w.compress {
```
Perhaps `w.rotateMu.Unlock()` here, so that we don't have to do the extra check in the `defer`?

cpuguy83 replied:
Updated, but more as we discussed on last weeks call.
This now sets a var for `noCompress := i w.maxFiles <= 1 || !w.compress`, then we check that var both in the defer and here.

Additionally added some docs around that in the defer.

thaJeztah reviewed (commented):
overall seems good; left one comment/suggestion (as discussed)

thaJeztah reviewed (approved):
LGTM if green

cpuguy83 commented:
All green!

StefanScherer reviewed (approved):
LGTM

tonistiigi reviewed (approved):
sgtm

tianon commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -167,56 +167,84 @@ func (w *LogFile) WriteLogEntry(msg *logger.Message) error {
+	} else {
+		var renameErr error
+		for i := 0; i < 10; i++ {
+			if renameErr = os.Rename(fname, fname+".1"); renameErr != nil && !os.IsNotExist(renameErr) {
```
If `fname.1` already exists, this will potentially be the same issue, right?  Should this use `i` or `i+1` instead?
(Sorry I'm not more familiar with this code to know whether this even makes sense :see_no_evil:)
